sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,623 @@
+            logDebug("[CBO] No range comparison statistics for String/Binary type " + attrRef)
+            None
+        }
+    }
+  }
+
+  /**
+   * This method converts a numeric or Literal value of numeric type to a BigDecimal value.
+   * In order to avoid type casting error such as Java int to Java long, we need to
+   * convert a numeric integer value to String, and then convert it to long,
+   * and then convert it to BigDecimal.
+   *
+   * @param attrDataType the column data type
+   * @param litValue the literal value
+   * @param litDataType the data type of literal value
+   * @return a BigDecimal value
+   */
+  def numericLiteralToBigDecimal(
+       attrDataType: DataType,
+       litValue: Any,
+       litDataType: DataType)
+    : BigDecimal = {
+    attrDataType match {
+      case _: IntegralType =>
+        BigDecimal(java.lang.Long.valueOf(litValue.toString))
+
+      case _: FractionalType =>
+        BigDecimal(litValue.toString.toDouble)
+
+      case DateType =>
+        val dateLiteral = litDataType match {
+          case StringType =>
+            DateTimeUtils.stringToDate(litValue.asInstanceOf[UTF8String])
+              .getOrElse(0).toString
+          case _ => litValue.toString
+        }
+        BigDecimal(java.lang.Long.valueOf(dateLiteral))
+
+      case TimestampType =>
+        val tsLiteral = litDataType match {
+          case StringType =>
+            DateTimeUtils.stringToTimestamp(litValue.asInstanceOf[UTF8String])
+                .getOrElse(0).toString
+          case _ => litValue.toString
+        }
+        BigDecimal(java.lang.Long.valueOf(tsLiteral))
+    }
+  }
+
+  /**
+   * Returns a percentage of rows meeting an equality (=) expression.
+   * This method evaluates the equality predicate for all data types.
+   *
+   * @param attrRef an AttributeReference (or a column)
+   * @param literal a literal value (or constant)
+   * @param update a boolean flag to specify if we need to update ColumnStat of a given column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   */
+  def evaluateEqualTo(
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Option[Double] = {
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+
+    // decide if the value is in [min, max] of the column.
+    // We currently don't store min/max for binary/string type.
+    // Hence, we assume it is in boundary for binary/string type.
+    val inBoundary: Boolean = attrRef.dataType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, attrRef.dataType).asInstanceOf[NumericRange]
+        val lit = numericLiteralToBigDecimal(attrRef.dataType, literal.value, literal.dataType)
+        (lit >= statsRange.min) && (lit <= statsRange.max)

```

Comment:
Can we define a method like `Range.inBoundary`? and then we can put the type matching stuff there.